### ISSUE TITLE
Errors when changing scene using change_scene_to_file()

### ISSUE THREAD

**Issue by golddotasksquestions** [godotengine/godot#85852]
### Tested versions

Reproducible in: 4.2 stable

### System information

Win 10, Godot 4.2 stable, Forward +, dedicated Nvidea RTX 1080

### Issue description

Using `get_tree().change_scene_to_file("res://some_scene.tscn")` works , but it produces a number of red errors. These seem to be different depending on the circumstances.

To test this, I made two scenes, **Scene A** and **Scene B**:
![image](https://github.com/godotengine/godot/assets/47016402/9e0fa0b0-6247-4e15-9814-0840628a3385)

--------------------------------------------------
In **Scene A** I change scene by walking into an Area3D with a CharacterBody3D (template script), triggering a `body_entered` signal:

```gdscript
extends Node3D

func _on_area_3d_body_entered(body):
	if "Player" in body.name:
		get_tree().change_scene_to_file("res://change_3d_scene_b.tscn")
```
This will produce these two errors:

> E 0:00:05:0401   change_3d_scene_A.gd:5 @ _on_area_3d_body_entered(): This function can't be used during the in/out signal.
>   <C++ Error>    Condition "locked" is true.
>   <C++ Source>   scene/3d/area_3d.cpp:289 @ _clear_monitoring()
>   <Stack Trace>  change_3d_scene_A.gd:5 @ _on_area_3d_body_entered()

> E 0:00:05:0401   change_3d_scene_A.gd:5 @ _on_area_3d_body_entered(): Removing a CollisionObject node during a physics callback is not allowed and will cause undesired behavior. Remove with call_deferred() instead.
>   <C++ Source>   scene/3d/collision_object_3d.cpp:113 @ _notification()
>   <Stack Trace>  change_3d_scene_A.gd:5 @ _on_area_3d_body_entered()

--------------------------------------------------

In **Scene B** I change scene by walking into an Area3D with a CharacterBody3D (template script), triggering a `body_entered` signal which will set a can_interact boolean variable in the character. If the player then presses an Input action, the character fires a global custom signal, which the level connects to and changes the scene:

Global Autoload:
```swift
signal change_level
```

Character:
```swift
var can_interact = false

func _physics_process(delta):
	if can_interact:
		if Input.is_action_just_pressed("ui_accept"):
			Global.change_level.emit()
```

Level:
```swift
extends Node3D

func _ready():
	Global.change_level.connect(change_to_other_level)

func _on_area_3d_body_entered(body):
	if "Player" in body.name:
		body.can_interact = true

func _on_area_3d_body_exited(body):
	if "Player" in body.name:
		body.can_interact = false

func change_to_other_level():
	get_tree().change_scene_to_file("res://change_3d_scene_a.tscn")
```
This will produce these four errors:

> E 0:00:07:0043   change_3d_scene_b.tscn::GDScript_7cg2h:34 @ _physics_process(): Condition "!is_inside_tree()" is true. Returning: Transform3D()
>   <C++ Source>   scene/3d/node_3d.cpp:343 @ get_global_transform()
>   <Stack Trace>  change_3d_scene_b.tscn::GDScript_7cg2h:34 @ _physics_process()

> E 0:00:07:0043   change_3d_scene_b.tscn::GDScript_7cg2h:34 @ _physics_process(): Condition "!is_inside_tree()" is true. Returning: Transform3D()
>   <C++ Source>   scene/3d/node_3d.cpp:343 @ get_global_transform()
>   <Stack Trace>  change_3d_scene_b.tscn::GDScript_7cg2h:34 @ _physics_process()

> E 0:00:07:0043   change_3d_scene_b.tscn::GDScript_7cg2h:34 @ _physics_process(): Parameter "body->get_space()" is null.
>   <C++ Source>   servers/physics_3d/godot_physics_server_3d.cpp:915 @ body_test_motion()
>   <Stack Trace>  change_3d_scene_b.tscn::GDScript_7cg2h:34 @ _physics_process()

> E 0:00:07:0043   change_3d_scene_b.tscn::GDScript_7cg2h:34 @ _physics_process(): Condition "!is_inside_tree()" is true. Returning: Transform3D()
>   <C++ Source>   scene/3d/node_3d.cpp:343 @ get_global_transform()
>   <Stack Trace>  change_3d_scene_b.tscn::GDScript_7cg2h:34 @ _physics_process()

Note: Line 34 in the character script is just `move_and_slide()`

--------------------------------------------------
Possibly related issue (however that's a different error): https://github.com/godotengine/godot/issues/84681

### Steps to reproduce

Use `get_tree().change_scene_to_file("res://some_scene.tscn")` as described above, or test the provided Minimal Reproduction Project below.

### Minimal reproduction project (MRP)

[change_scene_errors.zip](https://github.com/godotengine/godot/files/13588297/change_scene_errors.zip)

**Comment by akien-mga** [comment]
Both situations are caused by the change in 4.2 that makes `change_scene_to_*` methods no longer being deferred internally. When you call those methods, the current scene is deleted immediately, and the new scene is added to the root shortly after.

In this case you'd need to defer both calls to `change_scene_to_file` to work around that change and make it work similarly to what used to happen in 4.1.

**Comment by akien-mga** [comment]
> Where can I read up on the reasoning for this change?

It's discussed in https://godotengine.org/article/godot-4-2-arrives-in-style/#critical-and-breaking-changes with a link to the PR that changed it. There wasn't much discussion there at the time, as it was an attempt at fixing critical issues and we hadn't considered all implications. This was further discussed in #85251.

> It's also not documented in the [methods description](https://docs.godotengine.org/en/4.2/classes/class_scenetree.html#class-scenetree-method-change-scene-to-file):

See the note you quoted :)

>  Note: See change_scene_to_packed() for details on the order of operations.

**Comment by akien-mga** [comment]
It's not intended to only be called deferred. In most situations you can call it as is and have the scene change done right away.

The errors you get come from doing tree manipulation during a physics callback, as they mention. You would have the same issue trying to remove a physics node in `_on_area_3d_body_entered` - which is what changing the scene ends up doing too.

**Comment by AThousandShips** [comment]
You're not just removing that collision object when changing scenes though, so it's probably triggered by some of the other bodies, otherwise why would that warning be triggered?

**Comment by KoBeWi** [comment]
The physics callback comes from area, not body. If you do `get_parent().remove_child(self)` you should be getting the same error.

**Comment by golddotasksquestions** [comment]
> most method don't communciate their timing except when there's a companion method that doesn't defer part of the call.

Maybe this means `change_scene_to_file()` needs a companion method then? Similarly to `free()` and `queue_free()`? So maybe `queue_change_scene_to_file()`?

As a new user, I learned very early on to use `queue_free()` rather than `free()`, even though I did not fully understand the implications. A similarly named `queue_change_scene_to_file()` which executes deferred by default would be similarly intuitive, I guess. At least it would match existing syntax and would be more self-explanatory.

> Note that the method is equally confusing in the name in the previous behavior where it did all at the end of the frame, without any name indicating it wasn't instant.

I very much disagree. `get_tree().change_scene()` just worked. I can't recall it ever failing or produce any errors in all those years. Thanks to it, changing scenes was one of the least confusing aspects about Godot. Now with, not so much anymore.

>But I don't see how it's not transparent? It is documented and in the release post.

I'm telling you as a long time user: It's not. For you, for akien or other contributors and C++ magicians, what is written in the release nodes and API description might sound totally obvious and self-explanatory. I have read both the documentation and the release post many times before creating this issue. To me, a non C++ user who does not work with the Godot source every day: it's totally opaque, non transparent. Please accept this point of view.

**Comment by AThousandShips** [comment]
Then how would it be transparent? How would it be more transparent than writing exactly how it changed? Please explain how you'd like it to be explained to be cleared than that 🙂

**Comment by golddotasksquestions** [comment]
I don't understand the issue with how it worked before fully, I don't understand fully why the error appear. I don't think I'm in the position to explain something I don't fully understand myself.

[I thought I knew what's going on](https://www.reddit.com/r/godot/comments/18c01i6/comment/kc7v54c/?utm_source=share&utm_medium=web2x&context=3), but I'm obviously wrong. Which is why I opened this issue.

**Comment by AThousandShips** [comment]
I'm not sure why this wasn't noticed, perhaps people were being safe and calling this method deferred to not do scene tree changes during physics callbacks

But as I said, we're looking into adding specific clear errors, and the change was done with the goal to fix a number of bugs in the scene change sequence

**Comment by Lippanon** [comment]
> I don't understand what the problem was with `get_tree().change_scene()`, why it had to change...

The main relevant changes were done in https://github.com/godotengine/godot/pull/78988 and https://github.com/godotengine/godot/pull/85184, perhaps this will give you more context on the why.

> I don't understand fully why these errors appear now with the new method.

This was explained here https://github.com/godotengine/godot/issues/85852#issuecomment-1843293368 - specifically 'manipulation during a physics callback'.

Personally, I've always called this method deferred to avoid any potential problems. Usually games also have animations, and loading, instead of an immediate transition.
For what it's worth, a long existing demo (https://github.com/godotengine/godot-demo-projects/blob/master/loading/autoload/global.gd) explained how to make the change manually, with a deferred call for the reasons mentioned, that I used to use, in conjuction with threaded loading, to avoid errors with the old implementation of the methods in question.

**Comment by luiscesjr** [comment]
> Personally, I've always called this method deferred to avoid any potential problems. Usually games also have animations, and loading, instead of an immediate transition. For what it's worth, a long existing demo (https://github.com/godotengine/godot-demo-projects/blob/master/loading/autoload/global.gd) explained how to make the change manually, with a deferred call for the reasons mentioned, that I used to use, in conjunction with threaded loading, to avoid errors with the old implementation of the methods in question.

I too used to use deferred to load my scenes, but somehow the latest engine changes broke it. 
Anyways, the following code:

get_tree().change_scene_to_file.bind("res://some_scene.tscn").call_deferred()

Did not work for me, somehow it returned null, but what was null was the current scene? Or something like that, I forgot, but it looked like it was freeing before it was supposed to,  and not having anything on tree to change to, error happened. Apparently this is the new expected behavior, or so that's what I understood from all of it, sometimes I find confusing in my mind trying to translate English in my head.

Anyways, instead I used @Lippanon 's link, and that seems to be working here, with a little change on freeing the current scene:

```
	get_tree().current_scene.queue_free() # Instead of free()
	var packed_scene := ResourceLoader.load(path) as PackedScene
	var instanced_scene := packed_scene.instantiate()
	# Add it to the scene tree, as direct child of root
	get_tree().root.add_child(instanced_scene)
	# Set it as the current scene, only after it has been added to the tree
	get_tree().current_scene = instanced_scene
```

And while [the docs tell you what is going on](https://docs.godotengine.org/en/stable/tutorials/scripting/change_scenes_manually.html), the how to do it is a bit vague with lots of options to go to, but few examples. 

Strangely tough, in the docs it tells you to free() instead of queue_free(), which not only results in error by attempting to free a locked object (Not calling it deferred), but also seems unsafe.

**Comment by Lippanon** [comment]
> get_tree().change_scene_to_file.bind("res://some_scene.tscn").call_deferred()

I tried this in the MRP and it works with no errors, using v4.3.dev.custom_build [84692c625].
You can also do ```get_tree().call_deferred("change_scene_to_file", "res://change_3d_scene_a.tscn")``` and it also works, in both scene changes of the MRP, without any errors.

Notice that, in the demo above, the function that calls ```free()``` instead of ```queue_free()``` is being called deferred, which is why it says it's safe to do so.

As for the [docs](https://docs.godotengine.org/en/stable/tutorials/scripting/change_scenes_manually.html) you mention, they probably should mention 'deferred' somewhere, I agree.

**Comment by georgespatton** [comment]
What would be the code to utilize this new behavior correctly?  This print statement in this block returns null:

```
await get_tree().change_scene_to_file("res://scenes/new_area.tscn")
await get_tree().process_frame
print("CHANGED TO AREA:", get_tree().current_scene)
```

**Comment by freshlybakedhot** [comment]
I am a new user who is accustomed to using C #, so the correct code for this issue is:
`GetTree().CallDeferred("change_scene_to_file",path);`
And the original code was:
`GetTree().ChangeSceneToFile(path);`

This is a bit against habit

**Comment by Reintjuu** [comment]
The workaround works properly, but how do you get the result from the `change_scene_to_file` method when deferring the call? `call_deferred()` is documented to always return `null`. I'd like to receive and handle the errors returned by the method.

**Comment by MasterMann** [comment]
@freshlybakedhot your snippet no longer works in 4.2.1.
`_Ready' gets called on the new scene node, but `GetTree()` returns null, which seems like a bug.

**Comment by jahed** [comment]
It feels like change_scene_to_x functions are broken in 4.2 and @Lippanon's workaround is the way to go for now. I haven't encountered any issues with it so far and there doesn't seem to be any other stable solution.

In my case, I noticed that on 4.2.1, using change_scene_to_file does not delete the previous scene. It's still available but orphaned and signal handlers are still called. The docs say the scene is deleted "immediately" but that isn't happening and it's impossible from a user perspective to know why it isn't happening. I'd have to dig into engine code. I tried to queue_free the orphaned scene/node after the call_deferred but it closes the game without errors.

**Comment by fkeyzuwu** [comment]
These are some of the most common methods in Godot that almost every game would use. I don't think we should make the main way of using them feel like a workaround, especially since it worked fine for the most part beforehand.

**Comment by sutyrin** [comment]
Hey, +1 on this, and here is some input from a several-days-into-Godot noob. 

at physics process when I do 

```
		queue_free()
		get_tree().change_scene_to_file("res://menu.tscn")
``` 

It works, but once I let myself 

```
		await anim.animation_finished
``

Then scene changing line breaks and I have to rewrite as per recommended here: 

```
		get_tree().change_scene_to_file.bind("res://menu.tscn").call_deferred()
```

From very outside it occurs to me that upon using `await` for the first time the function magicalliy switches to async mode, which makes all other calls to be async (deferred) as well. A surprising experience.

Also from this useful conversation I take away a habit to change scenes asyncronously no matter what, thanks for that!

**Comment by DinoMC** [comment]
Same here. When pressing the "start" button on the main menu in my game (pressed signal), I call `get_tree().change_scene_to_packed(myscene) `

And it works, but also throws two "data.tree is null" errors. If I use call_deferred, it stills throws the errors.

This didn't use to happen, and I haven't changed anything in the main menu, so I have no idea what started it, but I can't find a way to get rid of the errors.
